File: InteractiveLive/Android/VideoOne/component/ToolKit/src/main/java/com/vertcdemo/core/net/http/HttpRequestHelper.java
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0

package com.vertcdemo.core.net.http;

import android.content.Context;
import android.util.Log;

import androidx.annotation.AnyThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import com.google.gson.JsonSyntaxException;
import com.vertcdemo.core.BuildConfig;
import com.vertcdemo.core.common.AppExecutors;
import com.vertcdemo.core.common.GsonUtils;
import com.vertcdemo.core.net.ServerResponse;
import com.vertcdemo.core.SolutionDataManager;
import com.vertcdemo.core.eventbus.AppTokenExpiredEvent;
import com.vertcdemo.core.eventbus.SolutionEventBus;
import com.vertcdemo.core.net.ErrorTool;
import com.vertcdemo.core.net.IRequestCallback;
import com.vertcdemo.core.utils.AppUtil;
import com.vertcdemo.core.R;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class HttpRequestHelper {
    private static final String TAG = "HttpRequestHelper";
    private static final String LOGIN_URL = BuildConfig.SERVER_URL + "/login";

    public static final OkHttpClient DEFAULT_OKHTTP_CLIENT = new OkHttpClient.Builder()
            .build();

    @AnyThread
    public static <T> void sendPost(JSONObject params,
                                    Class<T> resultClass,
                                    @NonNull IRequestCallback<ServerResponse<T>> callBack) {
        AppExecutors.networkIO().execute(() -> sendPost(LOGIN_URL, params, resultClass, callBack));
    }

    @WorkerThread
    public static <T> void sendPost(@NonNull String url,
                                    JSONObject params,
                                    @Nullable Class<T> resultClass,
                                    @NonNull IRequestCallback<ServerResponse<T>> callBack) {
        try {
            Context context = AppUtil.getApplicationContext();
            String language = context.getString(R.string.language_code);
            params.put("language", language);
        } catch (JSONException ignored) {

        }
        RequestBody body = RequestBody.create(MediaType.get("application/json; charset=utf-8"), params.toString());

        Request request = new Request.Builder()
                .url(url)
                .post(body)
                .build();
        Call call = DEFAULT_OKHTTP_CLIENT.newCall(request);

        try (final Response response = call.execute()) {
            Log.d(TAG, "Request: " + params);
            if (!response.isSuccessful()) {
                throw new IOException("http code = " + response.code());
            }

            final ResponseBody responseBody = response.body();
            if (responseBody == null) {
                throw new IOException("ResponseBody is null");
            }

            JSONObject json = new JSONObject(responseBody.string());
            Log.d(TAG, "Response: " + json);

            final int code = json.optInt("code");
            final String message = json.optString("message");

            if (code != 200) {
                if (code == ErrorTool.ERROR_CODE_TOKEN_EXPIRED || code == ErrorTool.ERROR_CODE_TOKEN_EMPTY) {
                    SolutionDataManager.ins().setToken("");
                    SolutionEventBus.post(new AppTokenExpiredEvent());
                }
                AppExecutors.mainThread().execute(() -> callBack.onError(code, message));
                return;
            }

            Object responseObject = json.opt("response");

            T data;
            if (responseObject == null
                    || resultClass == null
                    || resultClass == Void.class) {
                data = null;
            } else {
                data = GsonUtils.gson().fromJson(responseObject.toString(), resultClass);
            }

            final ServerResponse<T> sr = ServerResponse.create(code, message, data);
            AppExecutors.mainThread().execute(() -> callBack.onSuccess(sr));
        } catch (IOException | JsonSyntaxException | JSONException e) {
            Log.d(TAG, "post fail url:" + url, e);
            AppExecutors.mainThread().execute(() -> callBack.onError(ErrorTool.ERROR_CODE_UNKNOWN, e.getMessage()));
        }
    }
}

File: Client/Android/component/SolutionBase/src/main/java/com/vertcdemo/core/net/http/HttpRequestHelper.kt
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0
package com.vertcdemo.core.net.http

import android.content.Context
import android.util.Log
import androidx.annotation.AnyThread
import androidx.annotation.WorkerThread
import com.google.gson.reflect.TypeToken
import com.vertcdemo.base.BuildConfig
import com.vertcdemo.base.R
import com.vertcdemo.core.SolutionDataManager
import com.vertcdemo.core.common.AppExecutors.mainThread
import com.vertcdemo.core.common.AppExecutors.networkIO
import com.vertcdemo.core.common.GsonUtils.gson
import com.vertcdemo.core.eventbus.AppTokenExpiredEvent
import com.vertcdemo.core.eventbus.SolutionEventBus
import com.vertcdemo.core.net.IRequestCallback
import com.vertcdemo.core.net.ServerResponse
import com.vertcdemo.core.utils.AppUtil.applicationContext
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

object HttpRequestHelper {
    private const val TAG = "HttpRequestHelper"
    private const val ERROR_CODE_TOKEN_EXPIRED = 450
    private const val ERROR_CODE_TOKEN_EMPTY = 451
    private const val ERROR_CODE_UNKNOWN = -1

    private const val LOGIN_URL = BuildConfig.SERVER_URL + "/login"

    private val okHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(20, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS)
            .build()
    }

    @JvmStatic
    @AnyThread
    fun <T> sendPost(
        params: JSONObject,
        resultClass: Class<T>?,
        callBack: IRequestCallback<ServerResponse<T>?>
    ) {
        networkIO.execute { sendPost(LOGIN_URL, params, resultClass, callBack) }
    }

    @WorkerThread
    fun <T> sendPost(
        url: String,
        params: JSONObject,
        resultClass: Class<T>?,
        callBack: IRequestCallback<ServerResponse<T>?>
    ) {
        val context: Context = applicationContext
        val language = context.getString(R.string.language_code)
        params.put("language", language)

        val body = RequestBody.create(MediaType.get("application/json; charset=utf-8"), "$params")
        val request = Request.Builder()
            .url(url)
            .post(body)
            .build()
        val call = okHttpClient.newCall(request)
        try {
            call.execute().use { response ->
                Log.d(TAG, "Request: $params")
                if (!response.isSuccessful) {
                    throw IOException("http code = " + response.code())
                }
                val responseBody = response.body() ?: throw IOException("ResponseBody is null")
                val type = TypeToken.getParameterized(
                    ServerResponse::class.java, resultClass
                ).type
                val serverResponse = gson().fromJson<ServerResponse<T>>(responseBody.string(), type)
                val code = serverResponse.code
                if (code != 200) {
                    if (code == ERROR_CODE_TOKEN_EXPIRED || code == ERROR_CODE_TOKEN_EMPTY) {
                        SolutionDataManager.token = ""
                        SolutionEventBus.post(AppTokenExpiredEvent())
                    }
                    mainThread.execute { callBack.onError(code, serverResponse.message) }
                    return
                }
                mainThread.execute { callBack.onSuccess(serverResponse) }
            }
        } catch (e: IOException) {
            Log.d(TAG, "post fail url:$url", e)
            mainThread.execute { callBack.onError(ERROR_CODE_UNKNOWN, e.message) }
        }
    }
}
